File: src/hebrew_anagram/dictionary.py
"""Dictionary loading + preprocessing for Hebrew word lists."""

from __future__ import annotations

from pathlib import Path

from .letters import normalize_final_letters, remove_niqqud

_HEBREW_BLOCK_START = 0x0590
_HEBREW_BLOCK_END = 0x05FF


def _is_hebrew_only(s: str) -> bool:
    return all(_HEBREW_BLOCK_START <= ord(c) <= _HEBREW_BLOCK_END for c in s)


def load_words(
    path: str | Path,
    *,
    min_length: int = 2,
    drop_non_hebrew: bool = True,
    normalize_finals: bool = False,
    strip_niqqud: bool = True,
) -> list[str]:
    """Load and preprocess a Hebrew word list from disk.

    Reads *path* as UTF-8, applies the configured preprocessing steps to each
    line, and returns the deduplicated list of surviving words in first-seen
    order. The source file is never modified.
    """
    if not isinstance(path, (str, Path)):
        raise TypeError(f"path must be str or Path, got {type(path).__name__}")

    text = Path(path).read_text(encoding="utf-8")

    seen: set[str] = set()
    out: list[str] = []

    for line in text.splitlines():
        if not line.strip():
            continue
        if line.lstrip().startswith("#"):
            continue

        word = line.strip()
        if strip_niqqud:
            word = remove_niqqud(word)
        if normalize_finals:
            word = normalize_final_letters(word)

        if drop_non_hebrew and not _is_hebrew_only(word):
            continue
        if len(word) < min_length:
            continue
        if word in seen:
            continue

        seen.add(word)
        out.append(word)

    return out

File: web/src/lib/dictionary.ts
// Dictionary loading + matching.

import { isHebrewOnly, normalizeFinalLetters, removeNiqqud } from "./hebrew";
import { canFormWord } from "./matcher";

export interface PreprocessOptions {
  minLength?: number;
  dropNonHebrew?: boolean;
  normalizeFinals?: boolean;
  stripNiqqud?: boolean;
}

/**
 * Preprocess raw word-list text into a deduplicated array of words.
 *
 * Skips blank lines and `#` comments, strips niqqud (default on),
 * optionally normalizes final letters, drops non-Hebrew lines (default on),
 * filters by minimum length (default 2 — drops one-letter entries), and
 * deduplicates while preserving first-seen order.
 */
export function preprocessWordList(
  raw: string,
  {
    minLength = 2,
    dropNonHebrew = true,
    normalizeFinals = false,
    stripNiqqud = true,
  }: PreprocessOptions = {},
): string[] {
  const seen = new Set<string>();
  const out: string[] = [];

  for (const rawLine of raw.split(/\r?\n/)) {
    if (!rawLine.trim()) continue;
    if (rawLine.trimStart().startsWith("#")) continue;

    let word = rawLine.trim();
    if (stripNiqqud) word = removeNiqqud(word);
    if (normalizeFinals) word = normalizeFinalLetters(word);

    if (dropNonHebrew && !isHebrewOnly(word)) continue;
    if (word.length < minLength) continue;
    if (seen.has(word)) continue;

    seen.add(word);
    out.push(word);
  }

  return out;
}

export interface FindMatchingOptions {
  minLength?: number;
  normalizeFinals?: boolean;
  wildcard?: string;
}

/**
 * Filter a preprocessed word list down to entries formable from *letters*.
 * `letters` is passed through to `canFormWord` unchanged — pre-normalize
 * it yourself if `normalizeFinals` was used during preprocessing.
 */
export function findMatchingWords(
  letters: string,
  words: readonly string[],
  { wildcard = "?" }: { wildcard?: string } = {},
): string[] {
  return words.filter((w) => canFormWord(letters, w, { wildcard }));
}

/**
 * Fetch a UTF-8 word list from `url` and preprocess it.
 * Throws if the response is not OK.
 */
export async function loadWordsFromUrl(
  url: string,
  options: PreprocessOptions = {},
): Promise<string[]> {
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to load dictionary: HTTP ${res.status} ${res.statusText}`);
  }
  const text = await res.text();
  return preprocessWordList(text, options);
}
